### ISSUE TITLE
[Question]: What is an ideal mixer CLI for all users?

### ISSUE THREAD

**Issue by gtkramer** [clearlinux/mixer-tools#627]
Using the mixer CLI as an upstream is somewhat difficult due to the assumption built into mixer that the user is a downstream.  This difficulty is mostly encountered when trying to define the sources (bundles and repos) for mixer.  What's outlined below is just a brainstorm of what might help solve this problem.

```
mixer init
mixer init --format 1 --version 31290 --last-version 31280
```
Sets up workspace skeleton only with default format, version, and last version if not supplied.  These could live in mixer.state.  Uses values supplied otherwise from the command line.  Running this again increments the version if it exists.  Last version is incremented when `mixer build update` completes successfully.  No other work is done with this command.

```
mixer repo sources 31290 file:///home/releases/www/31290/clear/x86_64/os
mixer repo sources 31290,https://clearnas.ostc.intel.com file:///home/releases/www/31290/clear/x86_64/os
mixer repo sources file:///home/releases/www/31290/clear/x86_64/os
```
Creates any number of repo sources for mixer.  Sources are layered from left to right, with sources having highest priority on the left and the left priority on the right (priority in the dnf sense).  If a path is given that is not a valid repo, mixer copies the rpms into it's workspace and creates a repo automatically.  Here, users would not know about local-rpm or local-yum folder.  Using this command again wipes out any existing repo configurations in mixer's workspace.

```
mixer bundle sources 31290 /home/clrbuilder/projects/clr-bundles
mixer bundle sources 31290,https://github.com/clearlinux/clr-bundles /home/clrbuilder/projects/clr-bundles
mixer bundle sources /home/clrbuilder/projects/clr-bundles
mixer bundle sources <location>...
<location> = <git ref,git host>
<location> = <local filesystem>
```
Creates any number of bundle sources for mixer.  Sources are layered from left to right, where subsequent sources are overlayed on top of the previous ones.  A downstream would do `mixer bundle sources 31290 /home/clrbuilder/projects/clr-bundles/bundles` whereas an upstream would do `mixer bundle sources /home/clrbuilder/projects/clr-bundles/bundles`.  Here, users would not know about local-bundles folder.  Using this command wipes out mixbundles file and any internal tracking of the existing sources.

```
mixer bundle add --all | <list of bundles> | --default-bundles
mixer config set Swupd.CONTENTURL "${DISTRO_URL}/update"
mixer config set Swupd.VERSIONURL "${DISTRO_URL}/update"
```

Something like this above would be all an OSV would need to do in order to get going with mixer, a possible BKM of sorts.  There are a few goals that would be nice to achieve with a perhaps similar mixer CLI:
* Users do not need to be aware of the implementation of the mixer workspace, namely the local-rpms, local-yum, and local-bundles folders or the mixbundles file.  Code that DevOps needs to write to use mixer would no longer manage the mixer workspace, or at least manage it less.  Users come with their sources, and mixer consumes them.  However, this would require users to be minimally aware of their sources, which isn't necessarily a bad thing, and isn't much different from how it is today.
* Allows mixer to be used for any number of downstreams, fully customizable for wherever the sources are.  Can use default sources, or let users supply their own defaults when locations need to differ.
* The offline flag is no longer needed because it is up to the user to determine what is online or offline based on the sources they define.  Conditional behavior that happens during any mixer subcommand that keys of this offline flag is also no longer needed.
* The distinction between upstream and downstream goes away because the distinction is only relevant to how sources for mixer are defined, but that is now the responsibility of the user.  Documentation would probably need to keep this distinction to help explain the concept to users.  While it is an add for the user, it's not hard for the user to say, "I want bundles from 31290 first and then my bundles," since this information already needs to be supplied today and is what users typically understand.
* Mixer makes less assumptions for the user and puts the user in control of how they want to set up their sources.

I'm sure there are some more fine-grain details that I've overlooked given how the stateful nature of mixer makes this challenging, but I wanted to put the idea forward to see what other folks thought, and see if it might lead to a better mixer CLI experience for upstreams.

**Comment by gtkramer** [comment]
A CLI such as this would obsolete #626

**Comment by gtkramer** [comment]
Similarly, it would negate the need for a separate tool for upstream mixing vs downstream mixing since the differences are in how sources are defined.  There is a good deal of functionality built into a use case to let users start editing bundles with mixer, but this is probably good for a separate tool to align better to the unix philosophy.

**Comment by gtkramer** [comment]
Given there is quite a bit of behavior built into mixer that checks whether its inputs are defined correctly and sensibly, it may be useful to consolidate this into a subcommand that is optional for the user to run.  Something like `mixer validate`.  This would help to narrow the focus for each command to keep the command's functions consistent with what the user expects without any additional behavior.

**Comment by gtkramer** [comment]
A `mixer validate` command may be confusing with `mixer build validate`.  It might make sense to have `mixer validate workspace` and `mixer validate build` so validation operations are tied under a common subcommand of `validate`.

**Comment by gtkramer** [comment]
Ideally, the `mixer validate` commands are optional since they are not a required part of the process.  They would just help to give assurance that things are looking sensible for starting to mix and then after the mix finishes.

**Comment by gtkramer** [comment]
The `mixer validate` commands would be run either just before or after the `mixer build` commands.  Then, mixer does not attempt to perform validation as users are setting up their workspace.

**Comment by jwakre** [comment]
> ```
> mixer repo sources 31290 file:///home/releases/www/31290/clear/x86_64/os
> mixer repo sources 31290,https://clearnas.ostc.intel.com file:///home/releases/www/31290/clear/x86_64/os
> mixer repo sources file:///home/releases/www/31290/clear/x86_64/os
> mixer repo sources /home/workspaces/clear/packages/koji/results
> mixer repo sources /home/workspaces/clear/repo
> mixer repo sources <location>...
> <location> = build number,mirror host
> <location> = curlable path
> <location> = local file path to rpms outside mixer workspace (with or without repodata)
> ```
> 
> 
> Creates any number of repo sources for mixer.  Sources are layered from left to right, with sources having highest priority on the left and the lowest priority on the right (priority in the dnf sense).  If a path is given that is not a valid repo, mixer copies the rpms into it's workspace and creates a repo automatically.  Here, users would not know about local-rpm or local-yum folder.  Using this command again wipes out any existing repo configurations in mixer's workspace.  When it is run, mixer recreates the .yum-mix.conf again and its normalized repo workspace again.  Required to be run every time prior to starting a mix.  All sources must always be defined.

I agree that the CLI should be improved to make it easier to manage an upstream mix and that all repositories should be treated equally (no special treatment for upstream/local). Although, I disagree with repeatedly requiring all repositories to be set on the command line because it adds complexity that most downstream users would find challenging. Alternatively, I think it would be best to initialize the mix with separate default configurations for upstream and downstream mixes. For instance, when initializing an upstream mix, upstream related settings would not be configured and when initializing a downstream mix, a repo would be set that points to upstream. There is likely more work required to decouple Mixer's hard requirements on upstream/local concepts.
